### ISSUE TITLE
Duplicate HTTP headers

### ISSUE THREAD

**Issue by ruschestor** [pi-hole/pi-hole#5197]
### Versions

- Pi-hole version is v5.15.5 (Latest: v5.15.5)
- AdminLTE version is v5.18.4 (Latest: v5.18.4)
- FTL version is v5.21 (Latest: v5.21)

### Platform
- OS and version: Ubuntu 20.04.5 LTS
- Platform: NanoPi NEO3

### Expected behavior
Unique headers.

### Actual behavior / bug
The web server (lighttpd) response contains 7 duplicated headers:
```
X-Pi-hole: The Pi-hole Web interface is working!
X-Pi-hole: The Pi-hole Web interface is working!
X-Frame-Options: DENY
X-Frame-Options: DENY
X-XSS-Protection: 0
X-XSS-Protection: 0
X-Content-Type-Options: nosniff
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self' 'unsafe-inline';
Content-Security-Policy: default-src 'self' 'unsafe-inline';
X-Permitted-Cross-Domain-Policies: none
X-Permitted-Cross-Domain-Policies: none
Referrer-Policy: same-origin
Referrer-Policy: same-origin
```

### Steps to reproduce
1. Open "Network" tool in Web Developer (Firefox) / Developer tools (Chrome) console in a browser
2. Open http://[pihole-ip-address]/admin/
3. Pick "admin" (or login.php) page and look at the "Headers > Response headers" section.

### Screenshots
![PiHole-Duplicates-Headers](https://user-images.githubusercontent.com/48578125/223054960-12389033-8a67-4bbd-8390-15f66812c28f.png)

**Comment by rdwebdesign** [comment]
You change will fix the issue, but I'm trying to understand how this happened in the first place (I saw this problem before, but only one or two times).

Did you reinstall Pi-hole recently?

This error is happening because there is more than one config file setting the headers.

You didn't post a Debug Log, but I'm almost sure you have 2 config files adding PHP/fastcgi configuration. 
Check your `/etc/lighttpd/conf-enabled/` directory.

Please, let me know if you have this 2 files:
- `15-fastcgi-php.conf` was the old config, used by previous installs (and kept on updates).
- `15-pihole-admin.conf` is the new config file, installed on fresh installs.

If you have both files, the headers will be duplicated.

**Comment by dschaper** [comment]
Also I see your `lighttpd` version is different than mine. What OS are you running on?
![image](https://user-images.githubusercontent.com/2782850/223211812-58b186bd-af54-4353-9002-22f7a2ef5e6a.png)

**Comment by ruschestor** [comment]
> Did you reinstall Pi-hole recently?

No. I have been updating the Pi-Hole since 5.1.2


> You didn't post a Debug Log, but I'm almost sure you have 2 config files adding PHP/fastcgi configuration.
> Check your `/etc/lighttpd/conf-enabled/` directory.

Debug Log - https://tricorder.pi-hole.net/gFhSlcCN/
But I already changed "setenv.add-response-header" to "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file.

There are 5 files in "/etc/lighttpd/conf-enabled/" folder:
- `05-setenv.conf`
- `10-accesslog.conf`
- `10-fastcgi.conf`
- `15-fastcgi-php.conf`
- `15-pihole-admin.conf`


> Also I see your `lighttpd` version is different than mine. What OS are you running on?

Ubuntu 20.04.5 LTS

This is the latest version of "lighttpd" available to me.
```
# apt-cache policy lighttpd
lighttpd:
  Installed: 1.4.55-1ubuntu1.20.04.1
  Candidate: 1.4.55-1ubuntu1.20.04.2
  Version table:
     1.4.55-1ubuntu1.20.04.2 500
        500 http://ports.ubuntu.com focal-security/universe arm64 Packages
        500 http://ports.ubuntu.com focal-updates/universe arm64 Packages
 *** 1.4.55-1ubuntu1.20.04.1 100
        100 /var/lib/dpkg/status
     1.4.55-1ubuntu1 500
        500 http://ports.ubuntu.com focal/universe arm64 Packages
```

**Comment by ruschestor** [comment]
@rdwebdesign, @dschaper, what do you thing about this? I answered your questions.

**Comment by gstrauss** [comment]
What do you see with this? `grep -i X-Frame-Options -r /etc/lighttpd/`

Also, look at the output of `lighttpd -f /etc/lighttpd/lighttpd.conf -p` and see where (in which lighttpd config conditions) the headers are duplicated.  @rdwebdesign noted above, you might have a mix of old and new configs.

**Comment by dschaper** [comment]
> Please, let me know if you have this 2 files:
>
>    15-fastcgi-php.conf was the old config, used by previous installs (and kept on updates).
>    15-pihole-admin.conf is the new config file, installed on fresh installs.
>
> **If you have both files, the headers will be duplicated.**

You have both files in `/etc/lighttpd/conf-enabled`.

**Comment by ruschestor** [comment]
> What do you see with this? `grep -i X-Frame-Options -r /etc/lighttpd/`
> 
> Also, look at the output of `lighttpd -f /etc/lighttpd/lighttpd.conf -p` and see where (in which lighttpd config conditions) the headers are duplicated. @rdwebdesign noted above, you might have a mix of old and new configs.
I know that I have both files. I know how to fix this:

_To solve the issue it needs to change "setenv.add-response-header" lighttpd method to "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file._

There is my commit to fix the issue - (https://github.com/ruschestor/pi-hole/commit/2dc37f0ceed04940a43fe9c31253bf91767d38ab)

**Comment by yubiuser** [comment]
> To solve the issue it needs to change "setenv.add-response-header" lighttpd method to "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file.

I think the proper fix is to delete one of the files in `/etc/lighttpd/conf-enabled/`. As said above they should not be there at the same time.

**Comment by ruschestor** [comment]
> > To solve the issue it needs to change "setenv.add-response-header" lighttpd method to "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file.
> 
> I think the proper fix is to delete one of the files in `/etc/lighttpd/conf-enabled/`. As said above they should not be there at the same time.

But your sugпestion is manual task. I suggest fixing the issue for everyone with 1 change. I think that my issue is common.

**Comment by rdwebdesign** [comment]
> I think that my issue is common.

Not very common. 
We saw very few cases with similar problems.

> To solve the issue it needs to change "setenv.add-response-header" lighttpd method to "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file.

This works for you, but this is not a good fix for everyone.

Using `setenv.set-response-header` will overwrite any response header previously set and this is not desirable.

The whole idea behind the new `lighttpd` config was to avoid overwriting users config. Some users have custom configuration for Pi-hole and other users use `lighttpd` to serve different websites.

In your case, I think you have 2 different files setting the headers. You simply need to remove one of them.

We can try to find a permanent solution for your case.
Please generate a new debug log (the previous one is long gone).

**Comment by ruschestor** [comment]
I performed another research and found the core of issue. Let's go one by one.

@rdwebdesign 
**«15-fastcgi-php.conf was the old config, used by previous installs (and kept on updates).»
«In your case, I think you have 2 different files setting the headers. You simply need to remove one of them.»**
Deletion of «15-fastcgi-php.conf» doesn’t help. Headers still duplicates.


**«Using setenv.set-response-header will overwrite any response header previously set and this is not desirable.»**
I think you are mistaken.
The «setenv.set-response-header» directive replaces the value of a specific header.
At the same time the «setenv.add-response-header» directive from another config file and/or another «$HTTP["url"]» condition overwrites all previous headers.
You can check out the results of my research in the [blog](https://itgeeknotes.blogspot.com/2023/06/lighttpd-behavior-of-setenv-directives.html).

**«Some users have custom configuration for Pi-hole and other users use lighttpd to serve different websites.»**
All settings from the «15-pihole-admin.conf» config file applies to specific URL «/admin/» by condition «**$HTTP["url"] =~ "^/admin/"**». These settings can’t break other websites.

### The issue
As you can see in my comment of 07.03.2023 in the «/etc/lighttpd/conf-enabled/» folder I have 5 config files including «**05-setenv.conf**». This file contains single uncomment string: «**server.modules                += ( "mod_setenv" )**». But the «mod_setenv» has already enabled in «/etc/lighttpd/lighttpd.conf».
![05-setenv conf](https://github.com/pi-hole/pi-hole/assets/48578125/4bd3de65-b32e-4fb9-be8a-9919c7f6d278)
And after commenting of this line, the duplicate headers disappear.
I guess this issue/behavior was fixed because I couldn’t reproduce it on 1.4.63 version.

Therefore, I see 2 possible ways to solve the issue:
1. Replace "**setenv.add-response-header**" method by "**setenv.set-response-header**" in the "**/etc/lighttpd/conf-available/15-pihole-admin.conf**" config file.
2. Do not allow duplicate of enabling «**mod_setenv**» modules. Could you please find the right way (in initial installation and upgrade code) to fix this?

**Comment by ruschestor** [comment]
@rdwebdesign, Thanks.
The lighttpd version on my PiHole server - **1.4.55**.
The lighttpd version 1.4.63 is on my test VM with Ubuntu 22.04.2 LTS.

**Comment by rdwebdesign** [comment]
>The lighttpd version on my PiHole server - 1.4.55.

Hmmm... 
Testing with different versions can lead to different results (I'm not saying you should test again. There's no need).

---

This means the code to remove the duplicate config should have run when you updated the system, but apparently it failed on your system.
Not sure why, but that explains your problem.

**Comment by gstrauss** [comment]
> You can check out the results of my research in the [blog](https://itgeeknotes.blogspot.com/2023/06/lighttpd-behavior-of-setenv-directives.html).

@ruschestor FYI: lighttpd official documentation describes lighttpd behavior with directives in multiple config conditions, e.g. https://wiki.lighttpd.net/mod_setenv#Using-setenv-in-multiple-configuration-conditions

> Do not allow duplicate of enabling «mod_setenv» modules. Could you please find the right way (in initial installation and upgrade code) to fix this?

The duplication of `server.modules += ("mod_setenv")` should no longer be an issue on the modern, supported OS distro versions which ship lighttpd 1.4.56 or later, so I do not think pi-hole needs to make any changes with regards to `server.modules += ("mod_setenv")`.

> Replace "setenv.add-response-header" method by "setenv.set-response-header" in the "/etc/lighttpd/conf-available/15-pihole-admin.conf" config file.

lighttpd official documentation for [lighttpd mod_setenv](https://wiki.lighttpd.net/mod_setenv) states:
```
setenv.set-request-header (since 1.4.46)
setenv.set-response-header (since 1.4.46)
setenv.set-environment (since 1.4.46)
These directives set the given values, rather than appending the given values to the headers or environment.
These directives take precedence over the setenv.add-* counterparts.
```

@rdwebdesign: it would be safe to change /etc/lighttpd/conf-available/15-pihole-admin.conf" to use "setenv.set-response-header" under `/admin`.  The historical reason for using "setenv.add-response-header" was because "setenv.set-response-header" was not available in lighttpd before lighttpd 1.4.46.  As the version of lighttpd shipped by OS distro versions supported by pi-hole are much newer than lighttpd 1.4.46, pi-hole compatibility with lighttpd earlier than lighttpd 1.4.46 is no longer necessary, and "setenv.set-response-header" is the better choice, as suggested by @ruschestor 

@ruschestor wrote:
> The lighttpd version on my PiHole server - 1.4.55.

@ruschestor please consider upgrading your OS distro.  Most distros today, including RPi, offer a way to upgrade beteween major OS distro releases.

@rdwebdesign 
> This means the code to remove the duplicate config should have run when you updated the system, but apparently it failed on your system.
> Not sure why, but that explains your problem.

The pi-hole install script looks to avoid duplication of mod_setenv by looking in `lighttpd.conf` and in `15-pihole-admin.conf`.  According to @ruschestor, on their system there was an extra line in `05-setenv.conf` with `server.modules += ("mod_setenv")`, which is not handled by the install script.

The solution for @ruschestor with lighttpd 1.4.55 (and any lighttpd earlier than lighttpd 1.4.56) was that @ruschestor commented out the duplicated `server.modules += ("mod_setenv")` in `05-setenv.conf`.

Another solution for @ruschestor might be to uninstall pi-hole and make sure that the OS distro lighttpd package **original** lighttpd.conf was in /etc/lighttpd/lighttpd.conf (perhaps be removing and reinstalling the OS distro lighttpd package on their system), and then to reinstall pi-hole.

**Comment by ruschestor** [comment]
[gstrauss](https://github.com/gstrauss), many thanks for clarification!
[rdwebdesign](https://github.com/rdwebdesign), do you plan to fix the code? Thank you in advance!
